### PR TITLE
feat: Support Domain-Scoped Providers configuration for FeatureClient

### PR DESCRIPTION
**Pull request by arttonoyan**
This PR primarily focuses on adding support for Domain-Scoped Providers configuration, enabling improved management across different domains. Key changes include:

- Implementation of domain-scoped providers logic for flexible domain-based configurations.
- Addition of a `PolicyName` configuration to define the default `FeatureClient`.
- Introduction of an `AddProvider` extension method with default `OpenFeatureOptions` for streamlined provider setup.
- Refactoring and minor fixes to support build stability and clear naming conventions.

This enhancement provides a more adaptable and configurable environment for domain-scoped feature management.

---

#### Configuring LaunchDarkly as a New Provider

To configure LaunchDarkly as a new provider, begin by creating a `LaunchDarklyFactory`, a custom feature provider factory.

```csharp
public class LaunchDarklyFactory : IFeatureProviderFactory
{
    public Configuration Configuration { get; internal set; } = null!;

    public override FeatureProvider Build() => new Provider(Configuration);
}
```

#### Adding an Extension Method to OpenFeatureFactory

To streamline configuration, add an extension method, `AddLaunchDarkly`, to the `OpenFeatureBuilder`. This allows for either domain-scoped or default configurations:

**With Domain-Scoped:**

  ```csharp
  public static OpenFeatureBuilder AddLaunchDarkly(this OpenFeatureBuilder builder, string name, string stdKey, Action<ConfigurationBuilder>? configure = null)
  {
      builder.AddProvider<LaunchDarklyFactory>(name, builder => {
          var configBuilder = Configuration.Builder(stdKey);
          configure?.Invoke(configBuilder);
          builder.Configuration = configBuilder.Build();
      });
      return builder;
  }
  ```

**With default:**

  ```csharp
  public static OpenFeatureBuilder AddLaunchDarkly(this OpenFeatureBuilder builder, string stdKey, Action<ConfigurationBuilder>? configure = null)
  {
      builder.AddProvider<LaunchDarklyFactory>(builder => {
          var configBuilder = Configuration.Builder(stdKey);
          configure?.Invoke(configBuilder);
          builder.Configuration = configBuilder.Build();
      });
      return builder;
  }
  ```

#### Configuring OpenFeature with LaunchDarkly

Once the builder and extension methods are in place, configure `OpenFeature` to use LaunchDarkly with either a domain-scoped or default provider:

**Basic Configuration:**

  ```csharp
  builder.Services.AddOpenFeature(featureBuilder => {
      featureBuilder
          .AddHostedFeatureLifecycle()
          .AddContext(cfg => { })
          .AddLaunchDarkly(builder.Configuration["LaunchDarkly:SdkKey"]!); 
  });
  ```

**Domain-Scoped Provider Configuration:**

  ```csharp
  builder.Services.AddOpenFeature(featureBuilder => {
      featureBuilder
          .AddHostedFeatureLifecycle()
          .AddContext(cfg => { })
          .AddLaunchDarkly("anyName", builder.Configuration["LaunchDarkly:SdkKey"]!);
  });
  ```

#### Configuring Multiple Providers with Policy

To add multiple providers and configure a selection policy, define the default provider based on custom logic. This example sets `"anyName1"` as the default:

```csharp
builder.Services.AddOpenFeature(featureBuilder => {
    featureBuilder
        .AddHostedFeatureLifecycle()
        .AddContext(cfg => { })
        .AddLaunchDarkly("anyName1", builder.Configuration["LaunchDarkly:SdkKey_1"]!)
        .AddLaunchDarkly("anyName2", builder.Configuration["LaunchDarkly:SdkKey_2"]!)
        .AddPolicyName(options => {
            // Sets the default provider to "anyName1".
            // Custom logic can be implemented here to dynamically select a provider.
            options.DefaultNameSelector = serviceProvider => 
            {
                return "anyName1";
            };
        });
});
```

---

These examples provide a clear guide to integrating and configuring LaunchDarkly as a provider with flexible policy and lifecycle management, facilitating domain-scoped configurations in real-world applications.

### CODE REVIEW

**Comment by beeme1mr** [comment]
When you say "named provider", do you mean domain?

File: src/OpenFeature.DependencyInjection/IFeatureLifecycleManager.cs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-namespace OpenFeature;
+namespace OpenFeature.DependencyInjection;
```
Comment (toddbaert) [inline src/OpenFeature.DependencyInjection/IFeatureLifecycleManager.cs:1]: :+1:

**Comment by toddbaert** [comment]
Hey @arttonoyan I didn't quite have time to fully review this today, but it will be my priority tomorrow.

**Comment by arttonoyan** [comment]
> When you say "named provider", do you mean domain?



> When you say "named provider", do you mean domain?

**Comment by arttonoyan** [comment]
> When you say "named provider", do you mean domain?

Yes, I'll change the term to `domain` for clarity.

**Review by beeme1mr — APPROVED** [review]
I have limited experience with .NET but it looks good from an OpenFeature perspective. Thanks for your efforts here.

File: src/OpenFeature.DependencyInjection/OpenFeatureOptions.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    /// </summary>
+    public void AddDefaultProviderName() => AddProviderName(null);
+
+    /// <summary>
+    /// Registers a new feature provider name. This operation is thread-safe.
+    /// </summary>
+    /// <param name="name">The name of the feature provider to register. Registers as default if null.</param>
+    public void AddProviderName(string? name)
+    {
+        if (name == null)
```
Comment (askpt) [inline src/OpenFeature.DependencyInjection/OpenFeatureOptions.cs]: Does it make sense to check for white space or empty in this case?

File: src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs [hunk 3]
```diff
@@ -49,17 +52,216 @@ public static OpenFeatureBuilder AddContext(
...
+
+    /// <summary>
+    /// Adds a feature client to the service collection, configuring it to work with a specific context if provided.
+    /// </summary>
+    /// <param name="builder">The <see cref="OpenFeatureBuilder"/> instance.</param>
+    /// <param name="name">Optional: The name for the feature client instance.</param>
+    /// <returns>The <see cref="OpenFeatureBuilder"/> instance.</returns>
+    internal static OpenFeatureBuilder AddClient(this OpenFeatureBuilder builder, string? name = null)
+    {
+        if (name == null)
```
Comment (askpt) [inline src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs]: Does it make sense to check for empty or white space?

**Review by askpt — COMMENTED** [review]
I left two minor comments, but generally, it looks good.

File: src/OpenFeature.DependencyInjection/OpenFeatureOptions.cs (hunk 2 above)
Comment (arttonoyan) [reply src/OpenFeature.DependencyInjection/OpenFeatureOptions.cs]: Indeed, an additional check wouldn't hurt.

Comment (arttonoyan) [reply src/OpenFeature.DependencyInjection/OpenFeatureOptions.cs]: fixed

File: src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs (hunk 3 above)
Comment (arttonoyan) [reply src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs]: fixed

File: src/OpenFeature.DependencyInjection/OpenFeatureBuilder.cs [hunk 4]
```diff
@@ -17,4 +17,44 @@ public class OpenFeatureBuilder(IServiceCollection services)
...
+    /// </summary>
+    public bool IsPolicyConfigured { get; internal set; }
+
+    /// <summary>
+    /// Gets a value indicating whether a default provider has been registered.
+    /// </summary>
+    public bool HasDefaultProvider { get; internal set; }
+
+    /// <summary>
+    /// Gets the count of named feature providers that have been registered.
```
Comment (toddbaert) [inline src/OpenFeature.DependencyInjection/OpenFeatureBuilder.cs]:
```suggestion
    /// Gets the count of domain-bound providers that have been registered.
```

File: src/OpenFeature.DependencyInjection/OpenFeatureBuilder.cs [hunk 5]
```diff
@@ -17,4 +17,44 @@ public class OpenFeatureBuilder(IServiceCollection services)
...
+    /// This count does not include the default provider.
+    /// </summary>
+    public int NamedProviderRegistrationCount { get; internal set; }
```
Comment (toddbaert) [inline src/OpenFeature.DependencyInjection/OpenFeatureBuilder.cs]:
```suggestion
    public int DomainBoundProviderRegistrationCount { get; internal set; }
```

File: src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs [hunk 6]
```diff
@@ -49,17 +52,216 @@ public static OpenFeatureBuilder AddContext(
...
+    /// <typeparam name="TProviderFactory">The type of the provider factory implementing <see cref="IFeatureProviderFactory"/>.</typeparam>
+    /// <param name="builder">The <see cref="OpenFeatureBuilder"/> instance.</param>
+    /// <param name="configureFactory">An optional action to configure the provider factory of type <typeparamref name="TProviderFactory"/>.</param>
+    /// <returns>The configured <see cref="OpenFeatureBuilder"/> instance.</returns>
+    public static OpenFeatureBuilder AddProvider<TProviderFactory>(this OpenFeatureBuilder builder, Action<TProviderFactory>? configureFactory = null)
+        where TProviderFactory : class, IFeatureProviderFactory, new()
+        => AddProvider<OpenFeatureOptions, TProviderFactory>(builder, configureFactory);
+
+    /// <summary>
+    /// Adds a named feature provider with specified options and configuration builder.
```
Comment (toddbaert) [inline src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs]:
```suggestion
    /// Adds a feature provider with specified options and configuration builder for the specified domain.
```

File: src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs [hunk 7]
```diff
@@ -49,17 +52,216 @@ public static OpenFeatureBuilder AddContext(
...
+    /// </summary>
+    /// <typeparam name="TOptions">The <see cref="OpenFeatureOptions"/> type for configuring the feature provider.</typeparam>
+    /// <typeparam name="TProviderFactory">The type of the provider factory implementing <see cref="IFeatureProviderFactory"/>.</typeparam>
+    /// <param name="builder">The <see cref="OpenFeatureBuilder"/> instance.</param>
+    /// <param name="name">The unique name of the provider.</param>
```
Comment (toddbaert) [inline src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs]:
```suggestion
    /// <param name="domain">The unique name of the provider.</param>
```

File: src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs [hunk 8]
```diff
@@ -49,17 +52,216 @@ public static OpenFeatureBuilder AddContext(
...
+    /// <param name="configureFactory">An optional action to configure the provider factory of type <typeparamref name="TProviderFactory"/>.</param>
+    /// <returns>The <see cref="OpenFeatureBuilder"/> instance.</returns>
+    public static OpenFeatureBuilder AddProvider<TOptions, TProviderFactory>(this OpenFeatureBuilder builder, string name, Action<TProviderFactory>? configureFactory = null)
```
Comment (toddbaert) [inline src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs]:
```suggestion
    public static OpenFeatureBuilder AddProvider<TOptions, TProviderFactory>(this OpenFeatureBuilder builder, string domain, Action<TProviderFactory>? configureFactory = null)
```

File: src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs [hunk 9]
```diff
@@ -49,17 +52,216 @@ public static OpenFeatureBuilder AddContext(
...
+            return providerFactory.Create();
+        });
+
+        builder.AddClient(name);
+
+        return builder;
+    }
+
+    /// <summary>
+    /// Adds a named feature provider with a specified configuration builder, using default <see cref="OpenFeatureOptions"/>.
```
Comment (toddbaert) [inline src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs]:
```suggestion
    /// Adds a feature provider with a specified configuration builder for the specified domain, using default <see cref="OpenFeatureOptions"/>.
```

File: src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs [hunk 10]
```diff
@@ -49,17 +52,216 @@ public static OpenFeatureBuilder AddContext(
...
+    /// </summary>
+    /// <typeparam name="TProviderFactory">The type of the provider factory implementing <see cref="IFeatureProviderFactory"/>.</typeparam>
+    /// <param name="builder">The <see cref="OpenFeatureBuilder"/> instance.</param>
+    /// <param name="name">The unique name of the provider.</param>
```
Comment (toddbaert) [inline src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs]:
```suggestion
    /// <param name="domain">The unique domain of the provider.</param>
```

File: src/OpenFeature.DependencyInjection/MultiTarget/Guard.cs [hunk 11]
```diff
@@ -11,4 +11,10 @@ public static void ThrowIfNull(object? argument, [CallerArgumentExpression(nameo
         if (argument is null)
             throw new ArgumentNullException(paramName);
     }
+
```
Comment (toddbaert) [inline src/OpenFeature.DependencyInjection/MultiTarget/Guard.cs:14]: I can't comment on it, but I think this class should be in the DI namespace

File: src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs [hunk 12]
```diff
@@ -1,4 +1,7 @@
+using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Options;
+using OpenFeature.DependencyInjection;
 using OpenFeature.Model;
 
 namespace OpenFeature;
```
Comment (toddbaert) [inline src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs:7]: Should this be in the DI namespace?

File: src/OpenFeature.DependencyInjection/OpenFeatureServiceCollectionExtensions.cs [hunk 13]
```diff
@@ -1,53 +1,67 @@
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.DependencyInjection.Extensions;
-using OpenFeature.Internal;
-using OpenFeature.Model;
+using Microsoft.Extensions.Options;
+using OpenFeature.DependencyInjection;
+using OpenFeature.DependencyInjection.Internal;
 
 namespace OpenFeature;
```
Comment (toddbaert) [inline src/OpenFeature.DependencyInjection/OpenFeatureServiceCollectionExtensions.cs:7]: DI namespace?

**Review by toddbaert — APPROVED** [review]
LGTM in concept, I think the examples and code snippets you put in the description would be great in the README. I would add them, but instead of using LaunchDarkly, using the InMemory provider as your examples. The different domains example is especially helpful.

I've also found that quite a few classes are in `OpenFeature` instead of `OpenFeature.DI`, we might want to move them all there

**Review by lukas-reining — APPROVED** [review]
Looks good to me besides the DependencyInjection namespace @toddbaert brought up. :)

**Comment by arttonoyan** [comment]
> LGTM in concept, I think the examples and code snippets you put in the description would be great in the README. I would add them, but instead of using LaunchDarkly, using the InMemory provider as your examples. The different domains example is especially helpful.
> 
> I've also found that quite a few classes are in `OpenFeature` instead of `OpenFeature.DI`, we might want to move them all there

I’ve organized specific classes under the `OpenFeature.DependencyInjection` namespace within `OpenFeature.Hosting` projects to maintain consistent logical grouping across our packages. This approach aligns with best practices in .NET development, where unified namespaces enhance clarity and ease of configuration, especially as the project scales.

Using a single `OpenFeature` namespace in the `Program.cs` file also simplifies setup, allowing all configuration for OpenFeature and related dependencies to reside under a unified namespace. However, if you feel it’s best to keep namespaces closely aligned with project names, I’m open to making that change.

**Comment by arttonoyan** [comment]
@toddbaert Also, it might be a good idea to develop a `InMemoryProviderFactory` and an `AddInMemoryProvider` extension method within this package to make adding and using provider easier. This could also serve as an example of how other providers can be configured. What do you think, or would just writing a README be enough?

```
public class InMemoryProviderFactory : IFeatureProviderFactory
{
    internal IDictionary<string, Flag>? Flags { get; set; }

    public FeatureProvider Create() => new InMemoryProvider(Flags);
}

public static OpenFeatureBuilder AddInMemoryProvider(this OpenFeatureBuilder builder, string domain, Action<IDictionary<string, Flag>>? configure = null)
{
    builder.AddProvider<InMemoryProviderFactory>(domain, builder => {
        if (configure != null) {
            var flag = new Dictionary<string, Flag>();
            configure.Invoke(flag);
            builder.Flags = flag;
        }
    });

    return builder;
}
```

File: src/OpenFeature.DependencyInjection/MultiTarget/Guard.cs (hunk 11 above)
Comment (arttonoyan) [reply src/OpenFeature.DependencyInjection/MultiTarget/Guard.cs:14]: Done

**Comment by beeme1mr** [comment]
I think having an example of an in-memory provider extension and factory is a great idea. Thanks for adding that. It should make it easier for people to test and see how one can be created.

**Comment by arttonoyan** [comment]
> > LGTM in concept, I think the examples and code snippets you put in the description would be great in the README. I would add them, but instead of using LaunchDarkly, using the InMemory provider as your examples. The different domains example is especially helpful.
> > I've also found that quite a few classes are in `OpenFeature` instead of `OpenFeature.DI`, we might want to move them all there
> 
> I’ve organized specific classes under the `OpenFeature.DependencyInjection` namespace within `OpenFeature.Hosting` projects to maintain consistent logical grouping across our packages. This approach aligns with best practices in .NET development, where unified namespaces enhance clarity and ease of configuration, especially as the project scales.
> 
> Using a single `OpenFeature` namespace in the `Program.cs` file also simplifies setup, allowing all configuration for OpenFeature and related dependencies to reside under a unified namespace. However, if you feel it’s best to keep namespaces closely aligned with project names, I’m open to making that change.

@toddbaert @askpt What do you think, guys? Because I want to merge these changes.

**Comment by beeme1mr** [comment]
I think it's safe to merge this and get a final review on the SDK PR.

**Review by askpt — APPROVED** [review]
I did a final check, and all seems good. I will do another check as soon as this PR is merged into the SDK. :)

**Pull request merged**